### DISCUSSION TITLE
Additional reactions possible for createEntityAdapter?

### DISCUSSION THREAD

**Discussion by colorcube** [state-adapt/state-adapt#43] [Q&A] [answered]
What I want to do is something like this:

```typescript
  removeBooksOneBySecondId: (state, payload: string) => {
    const entities = booksAdapter.selectors.all(state.books);
    const entity = entities?.find(entity => entity.secondId == payload);
    if (entity) {
      labelsAdapter.removeOne(state.books, entity.id);
    }
  }
```

I don't see how to add that to createEntityAdapter adapter or with joinAdapters.

**Comment by mfp22** [comment, marked as answer]
There's a lot of logic in the entity adapter related to the `id`, so I don't think a 2nd ID would work. 

And because you don't know the `id` in advance, you can't create a filter selector that would automatically generate a `removeBooks<FilterSelectorName>` state change function. 

But `joinAdapters` lets you add more reactions. After the initial join adapters block, it's just like [`buildAdapter`](https://state-adapt.github.io/docs/core#buildadapter):

```typescript
adapter = joinAdapters<State>()({
  // adapters
})(([selectors, reactions]) => ({
 // more reactions
 removeBooksOneBySecondId: (state, payload: string) => {
    const entities = booksAdapter.selectors.all(state.books);
    const entity = entities?.find(entity => entity.secondId == payload);
    return !entity ? state : {
      ...state,
      books: labelsAdapter.removeOne(state.books, entity.id),
    };
  }
})();
```

And you can always just add the plain reaction function as part of a new adapter:

```typescript
newAdapter = createAdapter<State>()({
  ...oldAdapter,
  newStateChange: (state, payload: string) => {
    // ...
  }
});
```

`buildAdapter`/`joinAdapters` builder syntax is just syntactic sugar for this. 

But the most convenient way would be to use a group reaction inside `joinAdapters`/`buildAdapter`:

```typescript
adapter = joinAdapters<State>()({
  // adapters
})(([reactions, selectors) => ({
 // group reaction
 removeBooksOneBySecondId: {
    books: (state, payload: string) => { // Gets the `books` slice of the `state` object
      const entities = booksAdapter.selectors.all(books);
      const entity = entities?.find(entity => entity.secondId == payload);
      return !entity ? state : labelsAdapter.removeOne(state, entity.id); // Only need to return the `books` slice again
    },
  },
})();
```

**Comment by colorcube** [comment]
Thanks! I tried already group reaction but messed it up.
Here's the cleaned up code that compiles:

```typescript
  removeBooksOneBySecondId: {
    books: (state, payload: string) => { // Gets the `books` slice of the `state` object
      const entities = booksAdapter.selectors.all(state);
      const entity = entities?.find(entity => entity.name == payload);
      return !entity ? state : booksAdapter.removeOne(state, entity.id); // Only need to return the `books` slice again
    },
  },  
```

Means, `books:` is a "selector" for the state context of that adapter?

**Reply by mfp22** [reply]
Okay thanks, I edited my comment to match that.

`books` is referencing the actual state shape directly, not a selector.

**Reply by mfp22** [reply]
But yes, I think I know what you mean now. It's taking that state slice so the inner function doesn't have to worry about anything outside.
